application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,120 @@
+            @NotNull SlashCommandEvent event) {
+        return event.getJDA()
+            .openPrivateChannelById(userId)
+            .flatMap(channel -> channel.sendMessage(
+                    """
+                            Hey there, We are sorry to inform you that you have been warned for the following reason: %s
+                            The repercussion of this warning depends on the severity of the warning or one the number of warnings you have received.\040
+                            The warn system can be seen in the <#652440333107331082> channel.\040
+                            """
+                        .formatted(reason)))
+            .mapToResult()
+            .map(Result::isSuccess);
+    }
+
+    private static @NotNull MessageEmbed sendFeedback(boolean hasSentDm, @NotNull User target,
+            @NotNull Member author, @NotNull String reason) {
+        String dmNoticeText = "";
+        if (!hasSentDm) {
+            dmNoticeText = "(Unable to send them a DM.)";
+        }
+        return ModerationUtils.createActionResponse(author.getUser(), ModerationUtils.Action.WARN,
+                target, dmNoticeText, reason);
+    }
+
+    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+            @NotNull CharSequence reason, @NotNull Guild guild, @NotNull Interaction event) {
+
+        if (!ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event)) {
+            return false;
+        }
+        if (!ModerationUtils.handleHasBotPermissions(ACTION_VERB, Permission.BAN_MEMBERS, bot,
+                guild, event)) {
+            return false;
+        }
+        if (!ModerationUtils.handleHasAuthorPermissions(ACTION_VERB, Permission.BAN_MEMBERS, author,
+                guild, event)) {
+            return false;
+        }
+        return ModerationUtils.handleReason(reason, event);
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The option is null");
+        User target = userOption.getAsUser();
+        Member author = Objects.requireNonNull(event.getMember(), "The author is null");
+        Guild guild = Objects.requireNonNull(event.getGuild(), "The guild is null");
+        String reason = Objects.requireNonNull(event.getOption(REASON_OPTION), "The reason is null")
+            .getAsString();
+
+        Member bot = guild.getSelfMember();
+
+        if (!handleChecks(bot, author, reason, guild, event)) {
+            return;
+        }
+        long userId = target.getIdLong();
+        dmUser(userId, reason, event)
+            .map(hasSentDm -> sendFeedback(hasSentDm, target, author, reason))
+            .flatMap(event::replyEmbeds)
+            .queue();
+
+        actionsStore.addAction(guild.getIdLong(), author.getIdLong(), target.getIdLong(),
+                ModerationUtils.Action.WARN, null, reason);
```

Comment:
The order of this is wrong and you are also missing the log-message.

Look at `BanCommand`:
```java
sendDm(target, reason, guild, event)
  .flatMap(hasSentDm -> banUser(target, author, reason, deleteHistoryDays, guild)
  .map(banResult -> hasSentDm))
  .map(hasSentDm -> sendFeedback(hasSentDm, target, author, reason))
  .flatMap(event::replyEmbeds);
```
After sending the DM and before sending the feedback, you have to put the action into the store and write a log message, see `banUser` from `BanCommand`:
```java
logger.info(
        "'{}' ({}) banned the user '{}' ({}) from guild '{}' and deleted their message history of the last {} days, for reason '{}'.",
        author.getUser().getAsTag(), author.getId(), target.getAsTag(), target.getId(),
        guild.getName(), deleteHistoryDays, reason);

actionsStore.addAction(guild.getIdLong(), author.getIdLong(), target.getIdLong(),
        ModerationUtils.Action.BAN, null, reason);

return guild.ban(target, deleteHistoryDays, reason);
```